analyzers/src/SonarAnalyzer.CSharp/Rules/NewGuidShouldNotBeUsed.cs

```diff
@@ -30,30 +29,11 @@ namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     [Rule(DiagnosticId)]
-    public sealed class NewGuidShouldNotBeUsed : SonarDiagnosticAnalyzer
+    public sealed class NewGuidShouldNotBeUsed : NewGuidShouldNotBeUsedBase<SyntaxKind>
     {
-        internal const string DiagnosticId = "S4581";
-        private const string MessageFormat = "Use 'Guid.NewGuid()' or 'Guid.Empty' or add arguments to this Guid instantiation.";
+        protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
-
-        protected override void Initialize(SonarAnalysisContext context)
-        {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
-                {
-                    var objectCreationSyntax = (ObjectCreationExpressionSyntax)c.Node;
-
-                    if (objectCreationSyntax.ArgumentList?.Arguments.Count == 0 &&
-                        c.SemanticModel.GetSymbolInfo(objectCreationSyntax).Symbol is IMethodSymbol methodSymbol &&
-                        methodSymbol.ContainingType.Is(KnownType.System_Guid))
-                    {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, objectCreationSyntax.GetLocation()));
-                    }
-                },
-                SyntaxKind.ObjectCreationExpression);
-        }
+        protected override int? ConstructorArgumentListCount(SyntaxNode node) =>
+             (node as ObjectCreationExpressionSyntax)?.ArgumentList?.Arguments.Count;
```

Comment:
We should stick with the original approach with direct cast. We're registering for `SyntaxKind.ObjectCreationExpression` so there couldn't be anything else than `ObjectCreationExpressionSyntax`. The `?` part would not be coverable by tests.
```suggestion
            ((ObjectCreationExpressionSyntax)c.Node).ArgumentList?.Arguments.Count
```